engine/src/main/java/com/datatorrent/stram/ContainerRequestHandlerCloudera.java

```diff
@@ -0,0 +1,138 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.datatorrent.stram;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang3.tuple.MutablePair;
+import org.apache.hadoop.yarn.client.api.AMRMClient;
+import org.apache.hadoop.yarn.client.api.AMRMClient.ContainerRequest;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datatorrent.stram.StreamingContainerAgent.ContainerStartRequest;
+
+/**
+ * Handles creating container requests and issuing node-specific container
+ * requests by blacklisting specifically for cloudera
+ */
+public class ContainerRequestHandlerCloudera extends ContainerRequestHandler
+{
+  HashMap<ContainerRequest, ContainerStartRequest> hostSpecificRequests = new HashMap<>();
+  HashMap<ContainerRequest, ContainerStartRequest> otherContainerRequests = new HashMap<>();
+  HashMap<String, List<ContainerRequest>> hostSpecificRequestsMap = new HashMap<>();
+  List<String> blacklistedNodesForHostSpecificRequests = null;
+
+  public void reissueContainerRequests(AMRMClient<ContainerRequest> amRmClient, Map<StreamingContainerAgent.ContainerStartRequest, MutablePair<Integer, ContainerRequest>> requestedResources, int loopCounter, ResourceRequestHandler resourceRequestor, List<ContainerRequest> containerRequests, List<ContainerRequest> removedContainerRequests)
+  {
+    // Issue all host specific requests first
+    if (!hostSpecificRequestsMap.isEmpty() && requestedResources.isEmpty()) {
+      LOG.info("Issue Host specific requests first");
+      // Blacklist all the nodes and issue request for host specific
+      Entry<String, List<ContainerRequest>> set = hostSpecificRequestsMap.entrySet().iterator().next();
+      List<ContainerRequest> requests = set.getValue();
+      List<String> blacklistNodes = resourceRequestor.getNodesExceptHost(requests.get(0).getNodes());
+      amRmClient.updateBlacklist(blacklistNodes, requests.get(0).getNodes());
+      blacklistedNodesForHostSpecificRequests = blacklistNodes;
+      LOG.info("Sending {} request(s) after blacklisting nodes {} and removed host from request {}", requests.size(), blacklistNodes, requests.get(0).getNodes());
```

Comment:
Maybe it is better to log blacklisting all but ... since the list of nodes can be very large.